node.go

```diff
@@ -110,49 +136,134 @@ func MakeNode(nodeKey *NodeKey, buf []byte) (*Node, error) {
 		}
 		buf = buf[n:]
 
-		var (
-			leftNodeKey, rightNodeKey NodeKey
-			nonce                     int64
-		)
-		leftNodeKey.version, n, cause = encoding.DecodeVarint(buf)
+		mode, n, cause := encoding.DecodeVarint(buf)
 		if cause != nil {
-			return nil, fmt.Errorf("decoding node.leftNodeKey.version, %w", cause)
+			return nil, fmt.Errorf("decoding mode, %w", cause)
 		}
 		buf = buf[n:]
-		nonce, n, cause = encoding.DecodeVarint(buf)
-		if cause != nil {
-			return nil, fmt.Errorf("deocding node.leftNodeKey.nonce, %w", cause)
+		if mode < 0 || mode > 3 {
+			return nil, errors.New("invalid mode")
+		}
+
+		if mode&ModeLegacyLeftNode != 0 { // legacy leftNodeKey
+			node.leftNodeKey, n, cause = encoding.DecodeBytes(buf)
+			if cause != nil {
+				return nil, fmt.Errorf("decoding legacy node.leftNodeKey, %w", cause)
+			}
+			buf = buf[n:]
+		} else {
+			var (
+				leftNodeKey NodeKey
+				nonce       int64
+			)
+			leftNodeKey.version, n, cause = encoding.DecodeVarint(buf)
+			if cause != nil {
+				return nil, fmt.Errorf("decoding node.leftNodeKey.version, %w", cause)
+			}
+			buf = buf[n:]
+			nonce, n, cause = encoding.DecodeVarint(buf)
+			if cause != nil {
+				return nil, fmt.Errorf("deocding node.leftNodeKey.nonce, %w", cause)
+			}
+			buf = buf[n:]
+			if nonce < int64(math.MinInt32) || nonce > int64(math.MaxInt32) {
+				return nil, errors.New("invalid nonce, must be int32")
+			}
+			leftNodeKey.nonce = int32(nonce)
```

Comment:
A better test that avoids depending on the particular type of `nonce` is to convert back to the larger format:

```
leftNodeKey.nonce = int32(nonce)
if nonce != int64(leftNodeKey.nonce) {
    return nil, errors.New("nonce out of range")
}
```